### ISSUE TITLE
Minibatch processing

### ISSUE THREAD

**Issue by djsaunde** [Hananel-Hazan/bindsnet#245]
Minibatching data in artificial neural network training / inference allows for vast speedups in computation time and reductions in variance of parameter updates. Minibatching in spiking neural networks is inherently more complicated because spiking neurons maintain state (voltages, spikes traces, etc.).

We may be able to implement mini-batching by duplicating neurons at the start of each simulation, processing minibatches of data independently in parallel, and aggregating the results of each independent simulation to make, e.g., weight updates. The process would work something like this:

1. When calling `network.run`, users will pass in data of shape `[batch_size, time, k_1, ..., k_n]` (matching the WIP PR #243).
2. Neurons' state variables will be duplicated along the batch dimension `batch_size` times; when `batch_size = 1`, this reduces to our current implementation.
3. Simulation proceeds as normal, but updates to neuron state variables occur independently across the batch dimension. Updates to weights occur in parallel.

**Comment by djsaunde** [comment]
@k-chaney This may interest you.

**Comment by k-chaney** [comment]
This looks very interesting. I was thinking about this a little bit already. I need to read the `bindsnet.learning` module more closely to figure out what has to change there. Are there any other sticking points?

**Comment by djsaunde** [comment]
Hard to say. I think it will be relatively straightforward to implement, especially since we can borrow from PyTorch.

I think there are some conceptual issues:

- How to update adaptive neuron parameters, e.g., neurons' adaptive thresholds? Take average or sum of updates across batch dimension (average makes intuitive sense to me)?
- If we run two simulations back-to-back without resetting in between, how do we reconcile neurons' different initial conditions? Maybe enforce resetting between simulations?
- What do we do when batch sizes change between simulations?
- Should we adjust learning rates with batch sizes? E.g., if we run 10 examples at once with a learning rate of 0.01, should each example contribute to overall weight updates with a learning rate of 0.001?

Anyways, we can get _something_ working, then iron out these types of details.

**Comment by djsaunde** [comment]
Update: I have a working prototype of this. There are some bugs and design issues to figure out, but I have a simple example working.

**Comment by djsaunde** [comment]
Work is in progress on https://github.com/Hananel-Hazan/bindsnet/tree/batching.

**Comment by k-chaney** [comment]
I'll look into this branch. I just got back from CVPR which occupied my time completely last week.

**Comment by djsaunde** [comment]
Cool, I've implemented `batch_eth_mnist.py` and `batch_conv_mnist.py` as simple examples. Otherwise, there's some work left to do to fix some bugs.
